### PR TITLE
Track server metrics for all endpoints, remove high cardinality path label

### LINKED ISSUE #5025: [BUG] Service metrics endpoint excludes many important routes
### Willingness to contribute
The MLflow Community encourages bug fix contributions. Would you or another member of your organization be willing to contribute a fix for this bug to the MLflow code base?

- [x] Yes. I can contribute a fix for this bug independently.
- [ ] Yes. I would be willing to contribute a fix for this bug with guidance from the MLflow community.
- [ ] No. I cannot contribute a bug fix at this time.

### System information
- **Have I written custom code (as opposed to using a stock example script provided in MLflow)**: no
- **OS Platform and Distribution (e.g., Linux Ubuntu 16.04)**: Debian GNU/Linux 11
- **MLflow installed from (source or binary)**: binary
- **MLflow version (run ``mlflow --version``)**: 1.21.0
- **Python version**: 3.8
- **npm version, if running the dev UI**:
- **Exact command to reproduce**: see section below

### Describe the problem

The mlflow server option to expose a Prometheus metrics endpoint is a great observability feature for MLflow. Unfortunately, the current implementation leaves an incomplete view of the server health/performance. Currently, mlflow only logs metrics to a [subset of endpoints](https://github.com/mlflow/mlflow/blob/master/mlflow/server/prometheus_exporter.py#L18).

As of MLflow version 1.21, the following routes are not being included in the service metrics:

```
['static', '_get_experiment_by_name', '_create_experiment', '_list_experiments', '_get_experiment',
'_delete_experiment', '_restore_experiment', '_update_experiment',  '_update_run', '_delete_run', 
'_restore_run', '_set_experiment_tag',  '_delete_tag', '_get_run',  '_list_artifacts', '_get_metric_history',
'_log_batch', '_log_model', '_create_registered_model', '_rename_registered_model', 
'_update_registered_model',  '_delete_registered_model', '_get_registered_model', '_search_registered_models', 
'_list_registered_models',  '_get_latest_versions', '_create_model_version', '_update_model_version', 
'_transition_stage', '_delete_model_version',  '_get_model_version', '_search_model_versions', 
'_get_model_version_download_uri', '_set_registered_model_tag', '_set_model_version_tag', 
'_delete_registered_model_tag', '_delete_model_version_tag', 'health', 'serve_artifacts', 
'serve_model_version_artifact', 'serve_static_file', 'serve']
```

(see full list of endpoints)
```
from mlflow.server import app

app.view_functions.keys()
```

Filtering the set of routes to be included in the metrics endpoint seems like a potentially fragile approach as new routes are added in later versions of mlflow. It's especially problematic that the list of filtered routes cannot be configured. We currently have no way to monitor the health of the overall service given that many key routes (e.g. `log_batch`) are not included in the service metrics.

### Code to reproduce issue

Dockerfile for mlflow server
```
FROM python:3.8
RUN pip install mlflow==1.21.0

ENTRYPOINT mlflow server \
    --backend-store-uri sqlite:///mlflow.sqlite \
    --default-artifact-root file:///artifacts \
    --host 0.0.0.0 \
    --port 5000 \
    --expose-prometheus /prometheus
```

Build and run the Docker container
```
docker build -t mlflow_example -f Dockerfile .
docker run -p 5000:5000 mlflow_example
```

Script with incomplete representation in metrics endpoint
```
import mlflow
import random

mlflow.set_tracking_uri("http://127.0.0.1:5000")
mlflow.set_experiment("service_metrics")

with mlflow.start_run(run_name="test"):

    for _ in range(100):
        mlflow.log_metrics({
            'loss_a': random.random(),
            'loss_b': random.random(),
            'loss_c': random.random(),
        })

    mlflow.log_params({'a': 1, 'b': 2, 'c': 3})
```
See how metrics for these endpoints **_do not_** appear at http://127.0.0.1:5000/metrics

---

Script with expected representation in metrics endpoint
```
import mlflow
import random

mlflow.set_tracking_uri("http://127.0.0.1:5000")
mlflow.set_experiment("service_metrics")

with mlflow.start_run(run_name="test"):
    for _ in range(100):
        mlflow.log_metric('loss', random.random())

    mlflow.log_param('param', 'test')
```
See how metrics for these endpoints appear at http://127.0.0.1:5000/metrics

### Other info / logs
Include any logs or source code that would be helpful to diagnose the problem. If including tracebacks, please include the full traceback. Large logs and files should be attached.

### What component(s), interfaces, languages, and integrations does this bug affect?
Components 
- [ ] `area/artifacts`: Artifact stores and artifact logging
- [ ] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/docs`: MLflow documentation pages
- [ ] `area/examples`: Example code
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [ ] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/scoring`: MLflow Model server, model deployment tools, Spark UDFs
- [x] `area/server-infra`: MLflow Tracking server backend
- [ ] `area/tracking`: Tracking Service, tracking client APIs, autologging

Interface 
- [ ] `area/uiux`: Front-end, user experience, plotting, JavaScript, JavaScript dev server
- [ ] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [ ] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

Language 
- [ ] `language/r`: R APIs and clients
- [ ] `language/java`: Java APIs and clients
- [ ] `language/new`: Proposals for new client languages

Integrations
- [ ] `integrations/azure`: Azure and Azure ML integrations
- [ ] `integrations/sagemaker`: SageMaker integrations
- [ ] `integrations/databricks`: Databricks integrations

### PR DESCRIPTION
**Pull request by jeremyjordan**
## What changes are proposed in this pull request?

This PR updates the Prometheus metrics exporter to include all routes (except for the health endpoint) and removes the path label due to its high cardinality.

Fixes #5025

## How is this patch tested?

Added unit tests to ensure metrics are being exported with the appropriate labels.

## Does this PR change the documentation?

- [x] No. You can skip the rest of this section.
- [ ] Yes. Make sure the changed pages / sections render correctly by following the steps below.

1. Check the status of the `ci/circleci: build_doc` check. If it's successful, proceed to the
   next step, otherwise fix it.
2. Click `Details` on the right to open the job page of CircleCI.
3. Click the `Artifacts` tab.
4. Click `docs/build/html/index.html`.
5. Find the changed pages / sections and make sure they render correctly.

### Is this a user-facing change?

- [ ] No. You can skip the rest of this section.
- [x] Yes. Give a description of this change to be included in the release notes for MLflow users.

Prometheus metrics are now tracked for **_all_** endpoints; previously, metrics were only recorded for `["_search_runs", "_log_metric", "_log_param", "_set_tag", "_create_run"]`. Additionally, the metrics names are slightly different so you'll want to be sure to update your PromQL queries appropriately. `mlflow_requests_by_status_and_path_{bucket/count/sum}` metrics are now named `mlflow_http_request_duration_seconds_{bucket/count/sum}`. Finally, we no longer include the `path` label due to its high cardinality now that all endpoints (currently 46) are tracked.

### What component(s), interfaces, languages, and integrations does this PR affect?
Components 
- [ ] `area/artifacts`: Artifact stores and artifact logging
- [ ] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/docs`: MLflow documentation pages
- [ ] `area/examples`: Example code
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [ ] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/scoring`: MLflow Model server, model deployment tools, Spark UDFs
- [x] `area/server-infra`: MLflow Tracking server backend
- [ ] `area/tracking`: Tracking Service, tracking client APIs, autologging

Interface 
- [ ] `area/uiux`: Front-end, user experience, plotting, JavaScript, JavaScript dev server
- [ ] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [ ] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

Language 
- [ ] `language/r`: R APIs and clients
- [ ] `language/java`: Java APIs and clients
- [ ] `language/new`: Proposals for new client languages

Integrations
- [ ] `integrations/azure`: Azure and Azure ML integrations
- [ ] `integrations/sagemaker`: SageMaker integrations
- [ ] `integrations/databricks`: Databricks integrations

<a name="release-note-category"></a>
### How should the PR be classified in the release notes? Choose one:

- [x] `rn/breaking-change` - The PR will be mentioned in the "Breaking Changes" section
- [ ] `rn/none` - No description will be included. The PR will be mentioned only by the PR number in the "Small Bugfixes and Documentation Updates" section
- [ ] `rn/feature` - A new user-facing feature worth mentioning in the release notes
- [ ] `rn/bug-fix` - A user-facing bug fix worth mentioning in the release notes
- [ ] `rn/documentation` - A user-facing documentation change worth mentioning in the release notes

### CODE REVIEW

File: tests/server/test_prometheus_exporter.py [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+    from mlflow.server import app
+
+    ctx = app.app_context()
+    ctx.push()
+    yield app
+    ctx.pop()
+
+
+@mock.patch.dict(os.environ, {"PROMETHEUS_MULTIPROC_DIR": tmpdir.name})
+def test_metrics(app):
```
Comment (harupy) [inline tests/server/test_prometheus_exporter.py:23]: Can we use pytest's `tmpdir` fixture?

```suggestion
def test_metrics(app, tmpdir):
    with ock.patch.dict(os.environ, {"PROMETHEUS_MULTIPROC_DIR": tmpdir.strpath})
        ...
```

File: tests/server/test_prometheus_exporter.py [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+
+
+@pytest.fixture()
+def app():
+    from mlflow.server import app
+
+    ctx = app.app_context()
+    ctx.push()
+    yield app
+    ctx.pop()
```
Comment (harupy) [inline tests/server/test_prometheus_exporter.py:19]:
```suggestion
    with app.app_context():
        yield app
```

Can we use `with` to simplify the code?

File: mlflow/server/prometheus_exporter.py [hunk 3]
```diff
@@ -1,32 +1,19 @@
 from prometheus_flask_exporter.multiprocess import GunicornInternalPrometheusMetrics
 from flask import request
 
+from mlflow.version import VERSION
+
 
 def activate_prometheus_exporter(app):
-    metrics = GunicornInternalPrometheusMetrics(app, export_defaults=False)
+    def mlflow_version(req: request):  # pylint: disable=unused-argument
+        return VERSION
```
Comment (harupy) [inline mlflow/server/prometheus_exporter.py:9]:
```suggestion
    def mlflow_version(_: request):
        return VERSION

```

Can we use `_` here?

File: tests/server/test_prometheus_exporter.py [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+def test_metrics(app):
+    metrics = activate_prometheus_exporter(app)
+    metric_labels = {"method": "GET", "status": "200"}
+
+    with app.test_client() as c:
+        assert (
+            metrics.registry.get_sample_value("mlflow_http_request_total", labels=metric_labels)
+            is None
+        )
+        c.get("/")
```
Comment (harupy) [inline tests/server/test_prometheus_exporter.py:32]:
```suggestion
        resp = c.get("/")
        assert resp.status_code == 200
```

Can we assert the request is successful?

File: tests/server/test_prometheus_exporter.py [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+            metrics.registry.get_sample_value("mlflow_http_request_total", labels=metric_labels)
+            == 1
+        )
+
+        # calling the health endpoint should not increment the counter
+        c.get("/health")
+        assert (
+            metrics.registry.get_sample_value("mlflow_http_request_total", labels=metric_labels)
+            == 1
+        )
```
Comment (harupy) [inline tests/server/test_prometheus_exporter.py:50]:
```suggestion
        )
        resp = c.get("/api/2.0/mlflow/runs/get?run_id=123")
        assert resp.status_code == 404
        metric_labels = {"method": "GET", "status": str(resp.status_code)}
        assert (
            metrics.registry.get_sample_value("mlflow_http_request_total", labels=metric_labels)
            == 1
        )
```

Can we add a test case for the unsuccessful status code?

**Comment by harupy** [comment]
@jeremyjordan Thanks for the contribution, left some comments!

**Comment by jeremyjordan** [comment]
Thanks for the review @harupy! I've pushed a commit with updates based on your comments.

**Review by harupy — APPROVED** [review]
LGTM!

**Pull request merged**